Sources/KSCrashRecording/include/KSCrashAppMemory.h

```diff
@@ -0,0 +1,154 @@
+ */
+NS_ASSUME_NONNULL_BEGIN
+
+/** Notification sent when the memory level changes. */
+FOUNDATION_EXPORT NSNotificationName const KSCrashAppMemoryLevelChangedNotification;
+
+/** Notification sent when the memory pressure changes. */
+FOUNDATION_EXPORT NSNotificationName const KSCrashAppMemoryPressureChangedNotification;
+
+/** Notification keys that hold new and old values in the _userInfo_ dictionary. */
+FOUNDATION_EXPORT NSString *const KSCrashAppMemoryNewValueKey;
+FOUNDATION_EXPORT NSString *const KSCrashAppMemoryOldValueKey;
+
+/** The memory state for level and pressure. */
+typedef NS_ENUM(NSUInteger, KSCrashAppMemoryState) {
+
+    /** Everything is A-OK, go on with your business. */
+    KSCrashAppMemoryStateNormal = 0,
+
+    /** Things are starting to get heavy. */
+    KSCrashAppMemoryStateWarn,
+
+    /** Things are getting serious, allocations should be handled carefully. */
+    KSCrashAppMemoryStateUrgent,
+
+    /** At this point you are seconds away from being terminated.
+     *  You likely just received or are about to receive a
+     *  UIApplicationDidReceiveMemoryWarningNotification.
+     */
+    KSCrashAppMemoryStateCritical,
+
+    /** You have been or will be terminated. Out-Of-Memory. SIGKILL. */
+    KSCrashAppMemoryStateTerminal
+};
+
+/**
+ * AppMemory is a simple container object for everything important on Apple platforms
+ * surrounding memory.
+ */
+@interface KSCrashAppMemory : NSObject
+
+- (instancetype)init NS_UNAVAILABLE;
++ (instancetype)new NS_UNAVAILABLE;
+
+/** Footprint is the amount of memory used up against the memory limit (level). */
+@property(readonly, nonatomic, assign) uint64_t footprint;
+
+/**
+ * Remaining is how much memory is left before the app is terminated.
+ * same as `os_proc_available_memory`.
+ * https://developer.apple.com/documentation/os/3191911-os_proc_available_memory
+ */
+@property(readonly, nonatomic, assign) uint64_t remaining;
+
+/** The limi is the maximum amount of memory that can be used by this app,
+ *  it's the value that if attained the app will be terminated.
+ *  Do not cache this value as it can change at runtime (it's very very rare however).
+ */
+@property(readonly, nonatomic, assign) uint64_t limit;
+
+/** The current memory level. */
+@property(readonly, nonatomic, assign) KSCrashAppMemoryState level;
+
+/** The current memory pressure. */
+@property(readonly, nonatomic, assign) KSCrashAppMemoryState pressure;
+
+/** True when the app is totally out of memory. */
+- (BOOL)isOutOfMemory;
+
+@end
+
+/**
+ * Helpers to convert to and from pressure/level and strings.
+ * `KSCrashAppMemoryStateToString` returns a `const char*`
+ * because it needs to be async safe.
+ */
+FOUNDATION_EXPORT const char *KSCrashAppMemoryStateToString(KSCrashAppMemoryState state);
+FOUNDATION_EXPORT KSCrashAppMemoryState KSCrashAppMemoryStateFromString(NSString *const state);
+
+/**
+ * Internal and for tests.
+ */
+@interface KSCrashAppMemory ()
```

Comment:
I think leaving this interface in a public header might not make much of a difference since it would still be accessible to library users. A better approach might be to create a separate `.h` file, like `KSCrashAppMemory+Private.h`, and move it out of the `include` directory.